### ISSUE TITLE
Neo4j 3.0.4 embedded freezes when updating a legacy index

### ISSUE THREAD

**Issue by cmorgner** [neo4j/neo4j#7690]
**Neo4j version:** 3.0.4 Community
**Operating system:** Kubuntu Linux 16.04
**API/Driver:** Embedded Core API
### Expected behaviour

Create some nodes, add them to a legacy index, shut down the database and exit the program.
### Actual behaviour

Neo4j instance freezes and consumes 100% CPU time on a single core.
### Steps to reproduce

Run the following Java class in a Maven project with the below pom.xml.

``` java
public class Freeze {

    public static void main(final String[] args) {

        try {

            final File tmpDir                  = Files.createTempDirectory("neo4j").toFile();
            final GraphDatabaseService service = new GraphDatabaseFactory().newEmbeddedDatabase(tmpDir);

            try (final Transaction tx = service.beginTx()) {

                final Index<Node> index = service.index().forNodes("test");
                final Node node         = service.createNode();

                for (int i=0; i<20; i++) {

                    // create a new name for each iteration
                    final String name = "key" + i;

                    // remove node from index and add it again (update node with new value)
                    index.remove(node, name);
                    index.add(node, name, ValueContext.numeric(i));
                }

                tx.success();
            }

            service.shutdown();
            tmpDir.deleteOnExit();

        } catch (IOException ioex) {
            ioex.printStackTrace();
        }
    }
}
```

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.structr</groupId>
    <artifactId>neo4j-freeze-test</artifactId>
    <version>0.1-SNAPSHOT</version>
    <packaging>jar</packaging>
    <dependencies>
    <dependency>
        <groupId>org.neo4j</groupId>
        <artifactId>neo4j-community</artifactId>
        <version>3.0.4</version>
        <type>pom</type>
    </dependency>
    </dependencies>
    <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.source>1.7</maven.compiler.source>
    <maven.compiler.target>1.7</maven.compiler.target>
    </properties>
</project>
```
### Observations

The method `restoreNumericFields` of class `org.neo4j.index.impl.lucene.legacy.IndexType` seems to multiply the number of numeric `fields` in the Lucene `Document`.

Commenting out the line `index.remove(...)` in the above sample code fixes the issue (but prevents the previous value from being removed from the index, so subsequent index queries will return incorrect results).

**Comment by cmorgner** [comment]
**Update:** total execution time for the test code was 10m 51s, so not really a freeze but extreme slowdown.

The issue still exists in 3.1.0-M06

**Comment by jexp** [comment]
Do you have any profile info where it spent it's time?

Von meinem iPhone gesendet

**Comment by cmorgner** [comment]
![freeze](https://cloud.githubusercontent.com/assets/311577/17509703/9c73ed32-5e1b-11e6-87c1-1f96164cb7f6.png)

**Comment by jexp** [comment]
Thanks @cmorgner !!
@MishaDemianenko ^^^

**Comment by MishaDemianenko** [comment]
Thx @cmorgner and sorry for long reply time! 
Issue is fixed already by #8101 and released as part of 3.0.7.
Latest 3.1 includes it as well.

**Comment by cmorgner** [comment]
Confirming that the above test does not freeze Neo4j any more.
